### PR TITLE
#5362 GlusterFS fails to run more than once

### PR DESCRIPTION
**Pull request by ttindell2**
#5362  Added the ability to have the GlusterFS ansible script run more than once. It also allows to update the topology of the Gluster cluster even if the Gluster cluster has already been deployed.

@dustymabe

### CODE REVIEW

File: roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml [hunk 1]
```diff
@@ -189,14 +204,37 @@
...
+    glusterfs_heketi_client: "{% if glusterfs_heketi_is_native %}{{ openshift.common.client_binary }} rsh --namespace={{ glusterfs_namespace }} {%if not glusterfs_heketi_is_missing %}{{ heketi_pod.results.results[0]['items'][0]['metadata']['name'] }}{% else %}{{ deploy_heketi_pod.results.results[0]['items'][0]['metadata']['name'] }}{% endif %} {% endif %}{{ glusterfs_heketi_cli }} -s http://{% if glusterfs_heketi_is_native %}localhost:8080{% else %}{{ glusterfs_heketi_url }}:{{ glusterfs_heketi_port }}{% endif %} --user admin {% if glusterfs_heketi_admin_key is defined %}--secret '{{ glusterfs_heketi_admin_key }}'{% endif %}"
 
 - name: Verify heketi service
   command: "{{ glusterfs_heketi_client }} cluster list"
   changed_when: False
 
+- name: Place heketi topology on Heketi Pod
```
Comment (jarrpa) [inline roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml]: `s/Heketi/heketi/`

**Comment by jarrpa** [comment]
Also, can you squash your first two commits into one? Thanks. :)

File: roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml (hunk 1 above)
Comment (ttindell2) [reply roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml]: working on this now

File: roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml [hunk 2]
```diff
@@ -146,6 +146,21 @@
...
 
+- name: Get heketi admin secret
+  oc_secret:
+    state: list
+    namespace: "{{ glusterfs_namespace }}"
+    name: "heketi-{{ glusterfs_name }}-admin-secret"
+    decode: True
+  register: glusterfs_heketi_admin_secret
+
+- name: Set heteki admin key
```
[the remark this answers is not in the document]
Comment (ttindell2) [reply roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml]: working on this now

**Comment by ttindell2** [comment]
took out topology secret. Fixed the spelling errors

File: roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml [hunk 3]
```diff
@@ -189,14 +204,42 @@
...
+- name: Place heketi topology on heketi deploy Pod
+  shell: "{{ openshift.common.client_binary }} exec --namespace={{ glusterfs_namespace }} -i {{ deploy_heketi_pod.results.results[0]['items'][0]['metadata']['name'] }} -- bash -c 'mkdir -p {{ mktemp.stdout }} && cat > {{ mktemp.stdout }}/topology.json' < {{ mktemp.stdout }}/topology.json"
+  when:
+  - not glusterfs_heketi_deploy_is_missing
+
+- name: Place heketi topology on heketi Pod
+  shell: "{{ openshift.common.client_binary }} exec --namespace={{ glusterfs_namespace }} -i {{ heketi_pod.results.results[0]['items'][0]['metadata']['name'] }} -- bash -c 'mkdir -p {{ mktemp.stdout }} && cat > {{ mktemp.stdout }}/topology.json' < {{ mktemp.stdout }}/topology.json"
+  when:
+  - not glusterfs_heketi_is_missing
+
```
Comment (jarrpa) [inline roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml]: You don't need two tasks here, just use:
`{%if not glusterfs_heketi_is_missing %}{{ heketi_pod.results.results[0]['items'][0]['metadata']['name'] }}{% else %}{{ deploy_heketi_pod.results.results[0]['items'][0]['metadata']['name'] }}{% endif %}`

Comment (ttindell2) [reply roles/openshift_storage_glusterfs/tasks/glusterfs_common.yml]: made the change.

**Comment by jarrpa** [comment]
Code looks like it should work. I'll want to do some testing on it since the CI doesn't cover this yet. I'll try to get to it today.

**Comment by ttindell2** [comment]
Sounds good. I think changes should also be merged into 3.6. Do I need to make new pull request?

**Comment by jarrpa** [comment]
Yes, but let's wait for this one to merge first. I'd also want to have all the patches squashed into one. :)

**Comment by ttindell2** [comment]
Commits have been squashed. Fixed some issues with the topology load.

**Comment by sdodson** [comment]
Lint error needs to be resolved.
```
roles/openshift_storage_glusterfs/tasks/heketi_deploy_part1.yml:
  32:1      error    trailing spaces  (trailing-spaces)
yammlint issues found
ERROR: InvocationError: '/home/travis/build/openshift/openshift-ansible/.tox/py27-yamllint/bin/python setup.py yamllint'
```

**Comment by ttindell2** [comment]
Lint error has been fixed

**Comment by ttindell2** [comment]
Can someone rerun upgrade test? Test seems to have failed...

**Comment by sdodson** [comment]
/test upgrade


On Sep 20, 2017 8:24 PM, "Tim Tindell" <notifications@github.com> wrote:

Can someone rerun upgrade test? Test seems to have failed...

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/openshift/openshift-ansible/pull/5415#issuecomment-331016764>,
or mute the thread

**Comment by ttindell2** [comment]
Says it is missing LGTM though I see that it has been LGTMed

**Comment by ttindell2** [comment]
so retest tox? For whatever reason it is saying tox is not green even though it is...

**Comment by ttindell2** [comment]
ok now it is sayin the continuous integration isnt green even though it is....

**Comment by ttindell2** [comment]
It finally worked!!! 🥇

**Comment by ttindell2** [comment]
Ok so I am not sure why all the tests were rerun. They already were all passed a little bit ago. Can this get merged?

**Comment by ttindell2** [comment]
Can this be merged? The tests arleady passed once and are now failing for reasons that are not part of this merge...

**Pull request merged**